In [3]:
import numpy as np
import sklearn.datasets as ds

a = np.array(ds.load_iris().data)
b= a[:,1]

print(ds.load_iris().feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


# 1. Numpy Array Attributes
In this exercise we will use the `iris` dataset from the sklearn module. The features (independent variables) define properties of the flowers (see output above). The target (dependent variable) is the species of the flower. We stored the features in the array a.

 a. Get the number of dimensions of array a. Use the correct attribute\
 b. Get the shape of the array. How many rows and columns does the array have.\
 c. What datatype is used in the array?\

In [4]:
print(f'The number of dimensions is {a.ndim}. '
      f'The array has  {a.shape[0]} rows and {a.shape[1]} columns '
      f'with dtype {a.dtype}')


The number of dimensions is 2. The array has  150 rows and 4 columns with dtype float64


# 2. Indexing
Array b contains the second column of array a.\

a. From array b, get the 3th value (should be 3.2)\
b. From array b, get the 3th last value  (should be 3.0)\
c. From array a, get the value of the 3th column of the 91st row of array a (should be 4.4)\
d. From array a, from the last row get the first column (should be 5.9)\

In [5]:
#SOLUTION_START
#a
print(b[2])
#b
print(b[-3])
#c
print(a[90,2])
#d
print(a[-1,0])
#SOLUTION_END

3.2
3.0
4.4
5.9


## 3. Slicing
a. From array b, get the first 10 elements. (The features of the first 10 flowers.)\
b. From array b, get the flowers starting from the 10th row until the 20th row (including the 20th elemement)\
b. From array b, get all the flowers after the 50th row\
d. From array b, display the last 5 flowers\
e. From array b, display only every 3th flower\
f. From array b, display each 5th flower from the first 100 flowers\
g. From array a, display the rows of the first 10 elements (flowers)\
h. From array a, display the last two columns of the rows with index 10 to 20 (20 not included)\
i. From array a, display the second column of each second element.\
j. Based on array a, make a new array c with the 20 first rows with column 1 and 2. Make sure changes in the new array are not applied on the original aray.\
k. change element c[0,0] to 1. Check if a[0,0] is not changed\
l. Based on array a, make a new array l with 100 by 6 instead of 150 by 4.\
The outputs should look like this:
```
a=[3.5 3.  3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1]
b=[3.1 3.7 3.4 3.  3.  4.  4.4 3.9 3.5 3.8 3.8]
c=[3.2 3.2 3.1 2.3 2.8 2.8 3.3 2.4 2.9 2.7 2.  3.  2.2 2.9 2.9 3.1 3.  2.7
 2.2 2.5 3.2 2.8 2.5 2.8 2.9 3.  2.8 3.  2.9 2.6 2.4 2.4 2.7 2.7 3.  3.4
 3.1 2.3 3.  2.5 2.6 3.  2.6 2.3 2.7 3.  2.9 2.9 2.5 2.8 3.3 2.7 3.  2.9
 3.  3.  2.5 2.9 2.5 3.6 3.2 2.7 3.  2.5 2.8 3.2 3.  3.8 2.6 2.2 3.2 2.8
 2.8 2.7 3.3 3.2 2.8 3.  2.8 3.  2.8 3.8 2.8 2.8 2.6 3.  3.4 3.1 3.  3.1
 3.1 3.1 2.7 3.2 3.3 3.  2.5 3.  3.4 3. ]
d=[3.  2.5 3.  3.4 3. ]
e=[3.5 3.1 3.4 3.1 3.  4.4 3.8 3.7 3.4 3.5 3.1 4.2 3.5 3.4 3.2 3.  3.7 3.2
 2.8 2.4 2.  2.9 3.  2.5 2.5 3.  2.9 2.4 3.  2.3 2.6 2.3 2.9 2.8 3.  3.
 2.5 2.7 2.8 3.8 3.2 2.7 2.8 3.  2.8 3.  3.  3.1 3.3 3. ]
f=[3.6 3.1 4.  3.8 3.4 3.2 3.1 3.4 3.8 3.3 2.8 2.7 2.9 2.5 2.9 2.6 3.  2.5
 2.7 2.8]
g1=[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]
g2=[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]
h=[[1.5 0.2]
 [1.6 0.2]
 [1.4 0.1]
 [1.1 0.1]
 [1.2 0.2]
 [1.5 0.4]
 [1.3 0.4]
 [1.4 0.3]
 [1.7 0.3]
 [1.5 0.3]]
i=[3.5 3.2 3.6 3.4 2.9 3.7 3.  4.  3.9 3.8 3.4 3.6 3.4 3.4 3.4 3.1 4.1 3.1
 3.5 3.  3.5 3.2 3.8 3.8 3.7 3.2 3.1 2.8 3.3 2.9 2.  2.2 2.9 3.  2.2 3.2
 2.5 2.9 2.8 2.9 2.4 2.7 3.  3.1 3.  2.6 2.6 2.7 2.9 2.5 3.3 3.  3.  2.5
 2.5 3.2 3.  2.8 3.  2.6 3.2 2.8 3.3 2.8 2.8 2.8 2.8 2.6 3.4 3.  3.1 2.7
 3.3 2.5 3.4]
c[0,0]=1.0, a[0,0]=5.1
l. The shape of array l is (100, 6)
```


In [6]:
#SOLUTION_START
print(f'a={b[0:10]}') #Elements with index 0 to 9 (excluding the element with index 10)
print(f'b={b[9:20]}') #Elements with index 9 to 18
print(f'c={b[50:]}') #AFter 50th element = starting from index 50
print(f'd={b[-5:]}')
print(f'e={b[::3]}')
print(f'f={b[4:100:5]}') #The first 5th element has index. We take steps of 5 from there and we stop before index 100 (which is the 101th element)
print(f'g1={a[0:10,:]}')  #or shorter
print(f'g2={a[0:10]}')
print(f'h={a[10:20,-2:]}')
print(f'i={a[::2,1]}')
#j
j= a[:20,0:2].copy()
#k
j[0,0]=1
print(f'c[0,0]={j[0,0]}, a[0,0]={a[0,0]}')
#l
l = a.reshape((100,6))
print(f"The shape of array d is {l.shape}")
#m
#SOLUTION_START

a=[3.5 3.  3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1]
b=[3.1 3.7 3.4 3.  3.  4.  4.4 3.9 3.5 3.8 3.8]
c=[3.2 3.2 3.1 2.3 2.8 2.8 3.3 2.4 2.9 2.7 2.  3.  2.2 2.9 2.9 3.1 3.  2.7
 2.2 2.5 3.2 2.8 2.5 2.8 2.9 3.  2.8 3.  2.9 2.6 2.4 2.4 2.7 2.7 3.  3.4
 3.1 2.3 3.  2.5 2.6 3.  2.6 2.3 2.7 3.  2.9 2.9 2.5 2.8 3.3 2.7 3.  2.9
 3.  3.  2.5 2.9 2.5 3.6 3.2 2.7 3.  2.5 2.8 3.2 3.  3.8 2.6 2.2 3.2 2.8
 2.8 2.7 3.3 3.2 2.8 3.  2.8 3.  2.8 3.8 2.8 2.8 2.6 3.  3.4 3.1 3.  3.1
 3.1 3.1 2.7 3.2 3.3 3.  2.5 3.  3.4 3. ]
d=[3.  2.5 3.  3.4 3. ]
e=[3.5 3.1 3.4 3.1 3.  4.4 3.8 3.7 3.4 3.5 3.1 4.2 3.5 3.4 3.2 3.  3.7 3.2
 2.8 2.4 2.  2.9 3.  2.5 2.5 3.  2.9 2.4 3.  2.3 2.6 2.3 2.9 2.8 3.  3.
 2.5 2.7 2.8 3.8 3.2 2.7 2.8 3.  2.8 3.  3.  3.1 3.3 3. ]
f=[3.6 3.1 4.  3.8 3.4 3.2 3.1 3.4 3.8 3.3 2.8 2.7 2.9 2.5 2.9 2.6 3.  2.5
 2.7 2.8]
g1=[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1

## 4. Concatenation
a. Create a new array ca that concatenates the second column of array a with array b
b. Create a new array cb that concatenates the first 10 and last then rows of array a



In [7]:
#SOLUTION_START
a.dtype
b.dtype
ca = np.concatenate([a[:,0], b])
cb = np.concatenate([a[0:10], a[-10:]])
#SOLUTION_END